### ISSUE TITLE
no-await-in-loop rule occurs when an await is inside a do while loop

### ISSUE THREAD

**Issue by michaelseno** [eslint/eslint#10586]
Hi I am new in using ESLint. I've noticed that ESLint does have a rule `no-await-in-loop`. I'm kind of struggling to implement my logic due to the restriction of `await` inside a `loop`.  the code below will get `Promise {<pending>}` when i'm not appending the `await`. but works perfectly fine if it has `await` in it.

```javascript
async searchElement(locator) {
    let lookForElement;
    let limitCounter;

    do {
      lookForElement = await this.client.isVisible(locator); //this line will have the no-await-in-loop
      limitCounter = await this.client.isVisible(locator); //this line will have the no-await-in-loop

      if (lookForElement === false) {
        await this.swipe(81, 700, 81, 682); //this line will have the no-await-in-loop
        await this.pauseForInSeconds(0.7); //this line will have the no-await-in-loop
      } else {
        await this.clickOn(locator); //this line will have the no-await-in-loop
        break;
      }
    } while (limitCounter === false);
```
I believe that this code is valid but ESLint detecting all await should not be inside a loop. I would like to understand better on why there is such rule (I am very sorry I am very new to javascript and eslint). And since there is this rule. what are my other options to implement this logic since I really need to put it on a do while loop. Please let me know if you need more information. thanks alot.

P.S.
the temporary solution i'm doing now is to disable the rule by:
`// eslint-disable-next-line no-await-in-loop`

**Comment by ilyavolodin** [comment]
You can read more about this rule here: https://eslint.org/docs/rules/no-await-in-loop
Introduction does provide reasoning as to why you shouldn't use `await` inside loops. You are basically not using async aspect of the language correctly if you are using `await` inside the loop. That's why TC39 is introducing `for-await` construct (https://github.com/tc39/proposal-async-iteration). If you don't want to use that, you would want to push all of the function references into an array inside the loop, and then use `Promise.all` to execute all of those functions in async manner.

**Comment by nzakas** [comment]
It looks like this question has been answered. As this is a question rather than an action item, I'm closing the issue. If you still need help, please send a message to our [mailing list](https://groups.google.com/group/eslint) or [chatroom](https://gitter.im/eslint/eslint). Thanks!

**Comment by jtiscione** [comment]
This rule is assuming Promise.all() is always an option, which it usually isn't. I was surprised by this rule. Nobody's going to write a for loop with a bunch of async calls unless they need to run sequentially... right?

**Comment by vamshi9** [comment]
@ljharb I am sorry that I have to disagree with you! Because, if we have three state's and then if the second & third state is dependent on the update of first, there we cannot use `Promise.all` for sure because it messes up the sequential order.

**Comment by ljharb** [comment]
@vamshi9 you definitely still can :-) you can do something like:
```js
const a = doA();
const b = doB();
const c = b.then(() => doB());
return Promise.all([a, b, c]);
```
sequential order is preserved, maximum parallelism is achieved.

**Comment by vamshi9** [comment]
What if we have 1000's of state's @ljharb . My point is, it again disturbs the DRY principle of maybe calling some `do's`

**Comment by ljharb** [comment]
@vamshi9 This really isn't the place to discuss it, but the number of tasks you have isn't changed by using a loop or by using iterative means. Ping me on irc or in gitter with your code and I'll be happy to show you how to achieve it without loops.
